presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/BingTileFunctions.java

```diff
@@ -371,16 +393,111 @@ private static Point tileXYToLatitudeLongitude(int tileX, int tileY, int zoomLev
         return new Point(longitude, latitude);
     }
 
+    /**
+     * Returns a Bing tile at a given zoom level containing a point at a given latitude and longitude.
+     * Latitude must be within [-85.05112878, 85.05112878] range. Longitude must be within [-180, 180] range.
+     * Zoom levels from 1 to 23 are supported.
+     *
+     * @param latitude
+     * @param longitude
+     * @param zoomLevel
+     * @return
+     */
     private static BingTile latitudeLongitudeToTile(double latitude, double longitude, int zoomLevel)
+    {
+        long mapSize = mapSize(zoomLevel);
+        int tileX = longitudeToTileX(longitude, mapSize);
+        int tileY = longitudeToTileY(latitude, mapSize);
+        return BingTile.fromCoordinates(tileX, tileY, zoomLevel);
+    }
+
+    /**
+     * Called by bing_tiles_around, return immediate neighbour surrounding a point at a given latitude and longitude.
+     *
+     * @param latitude
+     * @param longitude
+     * @param zoomLevel
+     * @return
+     */
+    private static Set<BingTile> latitudeLongitudeToSurroundingTiles(double latitude, double longitude, int zoomLevel)
+    {
+        return latitudeLongitudeToSurroundingTiles(latitude, longitude, zoomLevel, 1);
+    }
+
+    /**
+     * Returns a collection of tiles at a given zoom level surrounding a point at a given latitude and longitude.
+     * It first locate the tile in the center.
+     * then it goes around to collect the tiles nearby
+     *
+     * @param latitude
+     * @param longitude
+     * @param zoomLevel
+     * @return
+     */
+    private static Set<BingTile> latitudeLongitudeToSurroundingTiles(double latitude, double longitude, int zoomLevel, int radius)
+    {
+        long mapSize = mapSize(zoomLevel);
+        long maxTileIndex = (mapSize / TILE_PIXELS) - 1;
+
+        checkCondition((radius >= 0 && radius <= maxTileIndex), "radius should be between 0 and map size");
+
+        int tileX = longitudeToTileX(longitude, mapSize);
+        int tileY = longitudeToTileY(latitude, mapSize);
+
+        Set<BingTile> bingTiles = new HashSet<>();
```

Comment:
You can use `ImmutableSet.Builder` to build an immutable set.

Reply:
This is exactly what I was thinking. 

The reason I added this comment is it's pretty common in the Presto codebase to use `ImmutableSet.Builder` when we build immutable sets.

Reply:
Do we need to use a `Set` here? If I understand correctly we can just use an `List` since `BingTile` with different coordinate is not equal to each other?

Reply:
@wenleix 
all BingTile are unique, and we honor the order among them, by definition we should go with Set in java.

> we can just use an List since BingTile with different coordinate is not equal to each other
what you mentioned is actually just the opposite reason of choosing List. Since we don't care about order, go with this stronger type will result in worse performance of appending, at least on paper.